### ISSUE TITLE
Paho doesn't properly deliver QOS0 messages all of the time

### ISSUE THREAD

**Issue by wes-johnson** [eclipse-paho/paho.mqtt.java#477]
Please fill out the form below before submitting, thank you!

- [ X] Bug exists Release Version 1.2.0 ( Master Branch)
- [ ] Bug exists in Snapshot Version 1.2.1-SNAPSHOT (Develop Branch)
- [ ] Bug exists in MQTTv5 Version (mqttv5-new Branch)

If this is a bug regarding the Android Service, please raise the bug here instead: https://github.com/eclipse/paho.mqtt.android/issues/new

In v1.2.0 Paho drops some messages especially when the publishes occur quickly.  The following test case fails in v1.2.0 but succeeds in v1.1.1.  This has been tested against Moquette and Mosquitto and the Paho client fails with either server when using v1.2.0.

```
    public void testInOrderDelivery() throws Exception {

        // Connect to the MQTT Server
        MqttConnectOptions options = new MqttConnectOptions();
        options.setAutomaticReconnect(false);
        options.setCleanSession(true);
        options.setConnectionTimeout(30);
        options.setKeepAliveInterval(30);
        options.setMaxInflight(500);
        options.setMqttVersion(MqttConnectOptions.MQTT_VERSION_3_1_1);

        MqttAsyncClient client = new MqttAsyncClient("tcp://localhost:1883", "PahoTestClientId");
        LocalCallback callback = new LocalCallback();
        client.setCallback(callback);
        client.connect(options);
        
        while (!client.isConnected()) {
        		Thread.sleep(100);
        }
        
        client.subscribe("1/2/3", 0);
        
        for (int i=0; i<10000; i++) {
        		client.publish("1/2/3", Integer.toString(i).getBytes(), 0, false);
        }
        
        while (!callback.isDone()) {
        		Thread.sleep(100);
        }

        	System.err.println("Success!");
    }
    
    private class LocalCallback implements MqttCallback {

    		private int messageIndex = 0;
    		private boolean done = false;
    		private ThreadPoolExecutor executor;

		public LocalCallback() {
			executor = new ThreadPoolExecutor(1, 1, 0L, TimeUnit.MILLISECONDS, new LinkedBlockingQueue<Runnable>());
		}
    		
    		public boolean isDone() {
    			return done;
    		}
    		
		@Override
		public void connectionLost(Throwable arg0) {
			// TODO Auto-generated method stub
		}

		@Override
		public void deliveryComplete(IMqttDeliveryToken arg0) {
			// TODO Auto-generated method stub
		}

		@Override
		public void messageArrived(String topic, final MqttMessage mqttMessage) throws Exception {
			executor.execute(new Runnable() {
				@Override
				public void run() {
					if (messageIndex != Integer.parseInt(new String(mqttMessage.getPayload()))) {
						System.err.println(
								"Expecting " + messageIndex + " but got " + new String(mqttMessage.getPayload()));
					}
					if (messageIndex == 9999) {
						done = true;
					} else {
						messageIndex++;
					}
				}
			});
		}
	}
```

**Comment by matsujun** [comment]
Maybe same issue as #421

**Comment by icraggs** [comment]
A fix has been added in 1.2.1 which should hopefully resolve this issue.

**Comment by icraggs** [comment]
Released in 1.2.1
